File: codes/rust/36.有效的数独.rs
/*
 * @lc app=leetcode.cn id=36 lang=rust
 *
 * [36] 有效的数独
 */

// @lc code=start
impl Solution {
    pub fn is_valid_sudoku(board: Vec<Vec<char>>) -> bool {
        let row_len = board.len();
        if row_len < 1 {
            return false;
        }
        let col_len = board[0].len();
        let mut row_flag: Vec<Vec<bool>> = vec![vec![false; 10]; row_len];
        let mut col_flag: Vec<Vec<bool>> = vec![vec![false; 10]; row_len];
        let mut box_flag: Vec<Vec<bool>> = vec![vec![false; 10]; row_len];
        for i in 0..row_len {
            for j in 0..col_len {
                if board[i][j] == '.' {
                    continue;
                }
                let num = board[i][j] as usize - '0' as usize;
                if row_flag[i][num] {
                    return false;
                }
                if col_flag[j][num] {
                    return false;
                }
                if box_flag[(i / 3) * 3 + j / 3][num] {
                    return false;
                }
                row_flag[i][num] = true;
                col_flag[j][num] = true;
                box_flag[(i / 3) * 3 + j / 3][num] = true;
            }
        }
        true
    }
}
// @lc code=end

File: codes/golang/36.有效的数独.go
/*
 * @lc app=leetcode.cn id=36 lang=golang
 *
 * [36] 有效的数独
 */

// @lc code=start
func isValidSudoku(board [][]byte) bool {
	rowbuf,colbuf,boxbuf := make([][]bool,9),make([][]bool,9),make([][]bool,9)
	for i:=0; i<9;i++ {
		rowbuf[i] = make([]bool,9)
		colbuf[i] = make([]bool,9)
		boxbuf[i] = make([]bool,9)
	}
	// 遍历一次，添加缓存
	for r:=0;r<9;r++ {
		for c:=0;c<9;c++ {
			if board[r][c] != '.' {
				num := board[r][c] - '0' - byte(1)
				if rowbuf[r][num] || colbuf[c][num] || boxbuf[r/3*3+c/3][num] {
					return false
				}
				rowbuf[r][num] = true
				colbuf[c][num] = true
				boxbuf[r/3*3+c/3][num] = true
			}
		}
	}
	return true
}
// @lc code=end

